Use external kubernetes authenticator in container based daemon.

thallgren opened the pull request:
## Description

This PR incorporates the authentication method introduced in PR #3052 into the flow when connecting to a container based daemon using `telepresence connect --docker`.

New `telepresence kubeauth-foreground --portfile <path>` subcommand.
This command starts the gRPC service that performs the host-based authentication. The server is started on demand the first time a `telepresence connect --docker` uses a kubernetes context with a user configured with an `exec` type authentication. When started, it allocates a port and announces that port as  a two byte binary integer in the given port file. The server will keep on running until it is killed or until the portfile is removed, and while running, it will touch the port file every three seconds so that stale port files can be detected and removed.

The port is picked up by the CLI binary and the address `docker.host.internal:<port>` is injected into a stubbed kubeconfig, which is then mounted into the docker container that runs the daemon. This approach makes it unnecessary to use environment variables to pass and address and a port.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [ ] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [ ] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

Commit:
Make it possible to pass system env when starting sub-process.

This is necessary because some k8s autenticators rely on specific
environment variables.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Pass address of k8s authentication service in the Exec stub.

This way, we don't have to pass any environment variables to the
authenticator process that runs in the container.

This commit also simplifies the patcher slightly so that it just
modifies the configuration without storing it, leaving that
responsibility to the caller.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add kubeauth-foreground subcommand to telepresence.

The `telepresence kubeauth-foreground` starts the new kubeconfig
authentication service daemon. It will run with the same environment
and kubernetes options as the caller process.

One advantage using the subcommand approach is that there's no extra
binary that must be downloaded and installed.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that daemons in docker use the kubeauth service when needed.

This commit adds the functionality needed to start the kubauth service
on the host on demand, and ensure that the docker daemon uses a patched
kubeconfig.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
The authenticator should not print usage when it errors.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add make rule that creates an authenticator that can run on the host.

This is for testing purposes only obviously, since the authenticator
is meant to run from within a container.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add linux specific --add-host

The `--add-host", "host.docker.internal:host-gateway` option is needed
in order to connect the `host.docker.internal` to the host's localhost.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
The patched authInfo.Exec can never run interactively.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

shepz reviewed (approved):
LGTM, but @knlambert could you take a look since you've worked on that part

Commit:
Ensure that the kubeauth server restarts on configuration changes

If the `KUBECONFIG` environment or `--kubeconfig` flag changes, or if
any of the files referenced by them changes, then the kubeauth server
must be restarted.

This commit ensures this restart by storing the list of configFiles
in the portfile that the server produces, and by adding a watcher for
both the port file and for the config files. The server exists if a
change is detected.

The client checks if the list of config files are the same, and if not,
it removes the portfile and restarts the server.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that container based daemon quits on early failure.

A container based daemon must quit immediately if the session creation
fails, because it can only handle one single session. A failure to quit,
means a failure to stop the container, which in turn means that the
container keeps on running and that the next connect attempt then fails
because the daemon container handling the given context already exists.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

knlambert commented on pkg/client/docker/kubeauth/cmd.go:
```diff
@@ -0,0 +1,185 @@
+}
+
+func (as *authService) watchFiles(ctx context.Context) error {
+	// If any of the files that the current kubeconfig uses changes, then we die
```
```suggestion
	// If any of the files that the current kubeconfig uses change, then we die
```

knlambert commented on pkg/client/docker/kubeauth/cmd.go:
```diff
@@ -0,0 +1,185 @@
+		return false
+	}
+	for dir := range dirs {
+		// Can't watch things that doesn't exist. We want to know if files in there change though.
```
```suggestion
		// Can't watch things that don't exist. We want to know if files in there change though.
```

knlambert commented on pkg/client/docker/kubeauth/cmd.go:
```diff
@@ -0,0 +1,185 @@
+	}
+}
+
+func (as *authService) watchFiles(ctx context.Context) error {
```
I'm not sure that I understand this, I assume it's to ensure that the kubeauth-foreground still uses the right kubeconfigs when we call a command ?

thallgren replied:
This is necessary because if the user adds, deletes, or modifies one of the files listed in the `KUBECONFIG` path, then a restart is necessary. A restart is also necessary if the client detects that the config has changed and removes the file.

knlambert reviewed (approved):
nitpicks but LGTM

Commit:
Fix grammar errors in comments.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
